doc/cprover-manual/goto-harness.md

```diff
@@ -0,0 +1,312 @@
+  nullpointers for generated values - in this case, this sets the _minimum_
+  length for our linked lists to one.
+* `nondet-globals` is non-deterministically initialising global variables.
+
+```
+CBMC version 5.11 (cbmc-5.11-1523-g419a958-dirty) 64-bit x86_64 linux
+[...]
+
+** Results:
+example.c function initialize_global
+[initialize_global.unwind.0] line 17 unwinding assertion loop 0: SUCCESS
+
+example.c function initialize_other
+[initialize_other.unwind.0] line 32 unwinding assertion loop 0: SUCCESS
+
+example.c function check_list
+[check_list.assertion.1] line 42 assertion list_parameter != global_linked_list: SUCCESS
+[check_list.unwind.0] line 50 unwinding assertion loop 0: SUCCESS
+[check_list.assertion.2] line 55 assertion global_cursor->value == parameter_cursor->value: FAILURE
+
+** 1 of 5 failed (2 iterations)
+VERIFICATION FAILED
+```
+
+We also have support for arrays (currently only for array function parameters,
+globals and struct members are considered for the future).
+
+Take this example of an implementation of an `is_prefix_of` function that
+should return true if the first string parameter `string` is a prefix of the
+second one `prefix`.
+
+```c
+// array_example.c
+
+int is_prefix_of(
+  const char *string,
+  int string_length,
+  const char *prefix,
+  int prefix_length
+)
+{
+  if(prefix_length > string_length) { return 0; }
+  // oops, we should have used prefix_length here
+  for(int i = 0; i < string_length; ++i)
+  {
+    // we'll get an out of bounds error here!
+    if(prefix[i] != string[i]) { return 0; }
+  }
+  return 1;
+}
+```
+
+We can compile and run it like this:
+
+```sh
+$ goto-cc -o array_example.gb array_example.c
+$ goto-harness \
+  --harness-function-name harness \
+  --harness-type call-function \
+  --function is_prefix_of \
+  --associated_array-size string:string_length \
+  --associated-array-size prefix:prefix_length \
+  array_example.gb array_example-mod.gb
+$ cbmc --function harness --pointer-check array_example-mod.gb
+```
+
+We have the additional option `associated-array-size` here. This is in the format
+`<array-parameter-name>:<array-size-parameter-name>` and will cause the array
+parameter with name`array-parameter-name` to be initialised as an array, with
+the parameter `array-size-parameter-name` holding its size (it should have
+some integral type like `int` or `size_t`).
+
+Running the example shows the bug highlighted in the comments:
+```
+[...]
+[is_prefix_of.pointer_dereference.6] line 14 dereference failure: pointer outside object bounds in prefix[(signed long int)i]: FAILURE
+```
+
+By default, arrays are created with a minimum size of 1 and a maximum size of 2.
+These limits can be set with the `--min-array-size` and `--max-array-size`
+options.
+
+If you have a function that takes an array parameter, but without an associated
+size parameter, you can also use the `--treat-pointer-as-array <parameter-name>`
+option.
```

Comment:
"but without an associated size parameter,"

"but the associated size parameter has unknown value"   ?